ipatests/test_webui/test_loginscreen.py

```diff
@@ -0,0 +1,409 @@
+        """
+        Log out and then log in using a new password.
+        It is need to check a new password
+        """
+        if (self.logged_in()):
+            self.logout()
+        self.load_url(self.get_base_url())
+        self.init_app(loginscreen.PKEY, loginscreen.PASSWD_ITEST_USER_NEW)
+        # User is not logged in
+        assert self.logged_in()
+
+    def reset_password(self, username=None, current_password=None,
+                       new_password=None, link_text=None, link_url=None):
+        """
+        Reset password with the given one
+        """
+        login_screen = self.get_login_screen()
+
+        if username is not None:
+            username_field = self.get_input_field('username', login_screen)
+        cur_pass_field = self.get_input_field('current_password', login_screen)
+        new_pass_field = self.get_input_field('new_password', login_screen)
+        verify_pass_field = self.get_input_field('verify_password',
+                                                 login_screen)
+        if username is not None:
+            username_field.send_keys(username)
+        cur_pass_field.send_keys(current_password)
+        new_pass_field.send_keys(new_password)
+        verify_pass_field.send_keys(new_password)
+        verify_pass_field.send_keys(Keys.RETURN)
+        self.wait()
+        self.assert_notification(assert_text='Password change complete',
+                                 link_text=link_text, link_url=link_url)
+
+    def get_data_from_form_row(self, form_row):
+        """
+        Parse data from the form record to a comparable structure
+        """
+        result = ()
+        label = self.find_xelement(".//label", form_row)
+        assert label.is_displayed()
+        result += (label.get_attribute('name'),)
+        result += (label.text,)
+        req = self.find_xelement("./../..", label)
+        result += (self.has_class(req, 'required'),)
+
+        field = self.find_xelement(".//input", form_row)
+        # not editable field
+        editable = field.is_displayed()
+        if not editable:
+            field = self.find_xelement(".//p", form_row)
+
+        result += (editable,)
+        assert field.is_displayed()
+        result += (field.get_attribute('type'),)
+        result += (field.get_attribute('name'),)
+        if not editable:
+            result += (field.text,)
+        else:
+            result += (field.get_attribute('value'),)
+        result += (field.get_attribute('placeholder'),)
+
+        return result
+
+    def get_data_from_button(self, button):
+        """
+        Parse data from the button to a comparable structure
+        """
+        result = ()
+        result += (button.get_attribute('name'),)
+        result += (button.get_attribute('title'),)
+
+        return result
+
+    def get_data_from_alert(self, data):
+        """
+        Parse data from the alert to a comparable structure
+        """
+        result = ()
```

Comment:
That was done because tuple is immutable type.
Increment here is to create new tuple contained old and appended one.
But for performance reason i changed it to list.

Reply:
I mean you could just write:
`return (data.text, )`
or ever simpler:
`assert (act_alert, ) == exp_alert`

Reply:
"get_data_from_form_row" and similar are used to gather data(as a tuple), which will be compared later against an expected one.

Reply:
so tuple here is a collection of data.text1, data.text2...

Reply:
Anyway, why do you create list, then add element to it and then convert it to a tuple?
Don't you find this operations redundant?
        `result = []`
        `result.append(data.text)`
        `return tuple(result)`

Reply:
I'm sorry. Maybe i don't understand you clearly.
For a calling convention "get_data_from_form_row" have to return the tuple.
The best approach i found is to create list, append to and then convert.

For example:
list_append.py
```python
#!/usr/bin/python3

def append_to_list():
    test_list = []
    for x in range(0, 50000):
        test_list.append(x)
    return tuple(test_list)

append_to_list()
```

```sh
time ./list_append.py 
0.04user 0.01system 0:00.06elapsed 96%CPU (0avgtext+0avgdata 10288maxresident)k
0inputs+0outputs (0major+1688minor)pagefaults 0swaps
```

tuple_append.py
```python
#!/usr/bin/python3

def append_to_tuple():
    test_tuple = ()  
    for x in range(0, 50000):
        test_tuple += (x,)
    return test_tuple

append_to_tuple()
```

```sh
time ./tuple_append.py 
5.38user 1.72system 0:07.11elapsed 100%CPU (0avgtext+0avgdata 10408maxresident)k
0inputs+0outputs (0major+1136956minor)pagefaults 0swaps
```

Reply:
If you have an idea, let me know, please.